lib/modules/versioning/hashicorp/convertor.spec.ts

```diff
@@ -0,0 +1,77 @@
+import { hashicorp2npm, npm2hashicorp } from './convertor';
+
+describe('modules/versioning/hashicorp/convertor', () => {
+  test.each`
+    constraint          | expected
+    ${'4.2.0'}          | ${'4.2.0'}
+    ${'4.2.0-alpha'}    | ${'4.2.0-alpha'}
+    ${'~> 4'}           | ${'>=4'}
+    ${'~> v4'}          | ${'>=4'}
+    ${'~> 4.0'}         | ${'^4.0'}
+    ${'~> 4.0.0'}       | ${'~4.0.0'}
+    ${'~> 4.0.1'}       | ${'~4.0.1'}
+    ${'~> 4.1'}         | ${'^4.1'}
+    ${'~> 4.1.0'}       | ${'~4.1.0'}
+    ${'~> 4.1.1'}       | ${'~4.1.1'}
+    ${'~> 4.0.0-alpha'} | ${'~4.0.0-alpha'}
+    ${'>= 4.0'}         | ${'>=4.0'}
+    ${'>= v4.0'}        | ${'>=4.0'}
+    ${'>=4.0'}          | ${'>=4.0'}
+    ${'<= 4.0'}         | ${'<=4.0'}
+    ${'<=4.0'}          | ${'<=4.0'}
+    ${'> 4.0'}          | ${'>4.0'}
+    ${'< 4.0'}          | ${'<4.0'}
+    ${'= 4.0'}          | ${'4.0'}
+    ${'> 4.0, < 5.0'}   | ${'>4.0 <5.0'}
+    ${'> 4.0,< 5.0'}    | ${'>4.0 <5.0'}
+    ${'~> 2.3.4'}       | ${'~2.3.4'}
+  `(
+    'hashicorp2npm("$constraint") === $expected',
+    ({ constraint, expected }) => {
+      expect(hashicorp2npm(constraint)).toBe(expected);
+    }
+  );
+
+  test('hashicorp2npm doesnt support !=', () => {
+    expect(() => hashicorp2npm('!= 4')).toThrow();
+  });
+
+  test('hashicorp2npm throws on invalid', () => {
+    expect(() => hashicorp2npm('^4')).toThrow();
+  });
+
+  test.each`
+    expected              | constraint
+    ${'4.2.0'}            | ${'4.2.0'}
+    ${'4.2.0-alpha'}      | ${'4.2.0-alpha'}
+    ${'4.0'}              | ${'4.0'}
+    ${'~> 4.0'}           | ${'^4'}
+    ${'~> 4.0'}           | ${'^4.0'}
+    ${'~> 4.0'}           | ${'^4.0.0'}
+    ${'~> 4.1'}           | ${'^4.1'}
+    ${'~> 4.1'}           | ${'^4.1.0'}
+    ${'~> 4.1, >= 4.1.1'} | ${'^4.1.1'}
+    ${'~> 4.0'}           | ${'~4'}
+    ${'~> 4.0.0'}         | ${'~4.0'}
+    ${'~> 4.0.0'}         | ${'~4.0.0'}
+    ${'~> 4.0.1'}         | ${'~4.0.1'}
+    ${'~> 4.1.0'}         | ${'~4.1.0'}
+    ${'~> 4.1.1'}         | ${'~4.1.1'}
+    ${'~> 4.0.0-alpha'}   | ${'~4.0.0-alpha'}
+    ${'>= 4.0'}           | ${'>=4.0'}
+    ${'<= 4.0'}           | ${'<=4.0'}
+    ${'> 4.0'}            | ${'>4.0'}
+    ${'< 4.0'}            | ${'<4.0'}
+    ${'> 4.0, < 5.0'}     | ${'>4.0 <5.0'}
+    ${'~> 2.3.4'}         | ${'~2.3.4'}
+  `(
+    'npm2hashicorp("$constraint") === $expected',
+    ({ expected, constraint }) => {
+      expect(npm2hashicorp(constraint)).toBe(expected);
```

Comment:
Is it intentional/deliberate to split the tests this way? Ie instead of having one table and verifying both directions in one go

Reply:
Didnt think of it, but actually yes, the cases are slightly different. but I like the idea of merging the ones that are the same, and i'll keep a separate block for the extra cases.

the extras are because there are cases that can be differently represented in the different languages